File: leetcode-easy/sqrtx/BS-solution.cpp
//
//  main.cpp
//  leetcode
//
//  Created by Macbook on 4/19/26.
//

#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    int mySqrt(int x) {
        int s = 1;
        int e = x;
        
        while (s <= e) {
            int y = s + (e - s) / 2;
            if (y > x/y) {
                e = y - 1;
            }
            else {
                s = y + 1;
            }
        }
        return e;
    }
};

int main(int argc, const char * argv[]) {
    // insert code here...
    Solution sl;
    int res = sl.mySqrt(17);

    cout << res << endl;
    return EXIT_SUCCESS;
}

File: leetcode-easy/sqrtx/BS-solution.py
class Solution:
    def my_sqrt(self, x: int) -> int:
        s: int = 1
        e: int = x
        while s <= e:
            y: int = s + (e - s) // 2
            if y > x // y:
                e = y - 1
            else:
                s = y + 1
        return e

if __name__ == "__main__":
    sol = Solution()
    print(sol.my_sqrt(8))
    print(sol.my_sqrt(5))
    print(sol.my_sqrt(3))
    print(sol.my_sqrt(1))
    print(sol.my_sqrt(0))
    print(sol.my_sqrt(10))